### PR TITLE
Restricted db-doc optimisation

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

Updates `db-doc` doc-level authorization to not rely on getting all `medic/contacts_by_depth` keys.
Instead, does a reverse authorization:
- runs `medic/docs_by_replication_key` view code over the context  docs (requested and stored) and collects all emitted keys
- collects all contact docs that correspond these keys
- runs `medic/contacts_by_depth` view code over the contacts, creating a "scoped" authorizationContext that only contains relevant allowed subjectIds (extracted from the contacts list view results)
- establish whether user has authorization using the "scoped" authorizationContext

medic/cht-core#5550

# Code review items

- Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
Testing performance with two users against a local deployment:

| User   | Load | Proposed (avg) | Master (avg) |
|--------|------|----------------|--------------|
| Medium | Idle | 14ms           | 39ms         |
| Medium | Load | 60ms           | 135ms        |
| Large  | Idle | 14ms           | 260ms        |
| Large  | Load | 60ms           | 820ms        |

User *medium* - has 1000 contacts, 9000 docs total
User *large* - has 8000 contacts, 18000 docs total

Idle = running a custom script against an idle database that requests docs the users can see. 
Load = both users replicating at the same time.

**Comment by dianabarsan** [comment]
@garethbowen mind having a look at this, please? Thanks!

File: api/src/services/db-doc.js [hunk 1]
```diff
@@ -82,29 +100,25 @@ module.exports = {
...
-        return storedDocs.filter(storedDoc => {
-          if (!storedDoc.ok) {
-            return false;
-          }
-
-          const viewResults = authorization.getViewResults(storedDoc.ok);
-          return authorization.allowedDoc(storedDoc.ok._id, authorizationContext, viewResults) ||
-                 authorization.isDeleteStub(storedDoc.ok);
-        });
+    let stored;
```
Comment (garethbowen) [inline api/src/services/db-doc.js]: Nit: I strongly prefer `const` over `let` and in this case it's only written to once so `const` is doable. In this case I think all you need to do is move the second `.then` into the first `.then` and the scope will work.

File: api/src/services/db-doc.js [hunk 2]
```diff
@@ -45,35 +56,42 @@ module.exports = {
   // this filters audited endpoints, so valid requests are allowed to pass through to the next route
   filterOfflineRequest: (userCtx, params, method, query, body) => {
     const isAttachment = params.attachmentId;
+    let stored;
+    let requested;
```
Comment (garethbowen) [inline api/src/services/db-doc.js]: Nit: as below - replace `let` with `const`. If it becomes too nested then feel free to pull out a function which takes these docs as params.

File: api/src/services/db-doc.js [hunk 3]
```diff
@@ -45,35 +56,42 @@ module.exports = {
...
-      .then(([ storedDoc, requestDoc, authorizationContext ]) => {
-        if (!storedDoc && !requestDoc) {
+      .then(([ storedDoc, requestDoc ]) => {
+        stored = getDocCtx(storedDoc);
+        requested = getDocCtx(requestDoc);
+
+        return authorization.getScopedAuthorizationContext(userCtx, [ stored, requested ]);
+      })
+      .then(authorizationContext => {
+        if (!stored && !requested) {
```
Comment (garethbowen) [inline api/src/services/db-doc.js]: Why `getScopedAuthorizationContext` when we know we're going to return false here? Can this be moved up to the previous block to return before the additional call?

File: api/src/services/db-doc.js (hunk 1 above)
Comment (dianabarsan) [reply api/src/services/db-doc.js]: Updated. I probably hate nesting promises more than `let`, but this didn't turn out so bad :)

**Pull request merged**